### ISSUE TITLE
efficient merges for Heaps

### ISSUE THREAD

**Issue by johnynek** [typelevel/cats-collections#165]
we can always merge two heaps by taking the smaller of the two and adding each element. This would be `n log n` where n is the size of the smaller heap. This totally discards the structure of the smaller one.

It *seems* like we should be able to leverage the heap property to more efficiently merge (say O(log n) would be ideal).

The original paper this is based on does not have a merge operation, but there may be other heap algorithms or small changes we can make to this implementation to support merge.

**Comment by johnynek** [comment]
I've researched this a bit. The Heap we have here is asymptotically optimal even though it is immutable for all the operations it supports.

The paper does not give a merge implementation. It may be possible to find an O(log N) implementation, but I don't yet know one and don't see it immediately.

Alternatively, we could have a second implementation of say `PairingHeap[A]` which while lacking a proof that it is efficient (open question for quite a while it seems) empirically performs very well, supporting log N merges:

https://www.schoolofhaskell.com/user/edwardk/revisiting-matrix-multiplication/part-5

We could have two implementations, `Heap` and `MergeableHeap` or `PairingHeap` or something, so if you know you don't need merges and want the very best performance, use Heap, if you need to merge use `PairingHeap`.

It would be cool to discover a merge algorithm on the current heap.

What do you think @larsrh ?

**Comment by vkostyukov** [comment]
There is also [`SkewHeap` and `LeftistHeap`](https://github.com/vkostyukov/scalacaster/tree/master/src/heap); both support fast, O(log n), merge.

I wrote that "standard heap" paper and I wasn't able to find an efficient merge algorithm either. I had some notes somewhere but I probably lost them at this point. What rings a bell is that, later, I stumbled upon old ML (language) lectures talking about a very similar approach to implementing heaps w/o mutations. I can't remember if they had merge implemented though.
